flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/BatchTableSourceScan.scala

```diff
@@ -38,7 +38,9 @@ class BatchTableSourceScan(
 
   override def deriveRowType() = {
     val flinkTypeFactory = cluster.getTypeFactory.asInstanceOf[FlinkTypeFactory]
-    flinkTypeFactory.buildRowDataType(tableSource.getFieldsNames, tableSource.getFieldTypes)
+    flinkTypeFactory.buildRowDataType(
+      tableSource.getFieldsNames,
+      TableEnvironment.getFieldTypes(tableSource.getReturnType))
```

Comment:
Should be fine if we add support for `AtomicType` in `TableEnvironment.getFieldTypes`